### ISSUE TITLE
Port qual, safe_repr and safe_str (from twisted.python.reflect) to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5913]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5913|
|Type|enhancement|
|Created|2012-08-24 18:03:46Z|
|Branch|https://github.com/twisted/twisted/tree/reflect-python3-5913-2|

`qual`, `safe_repr` and `safe_str` are used by `twisted.python.failure` and `twisted.python.log`, so it would be useful to port them to Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__5913 5913
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_reflect_python3_5913_2 branches/reflect-python3-5913-2
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1345831426000000 1345831426000000
changetime__1346440642000000 1346440642000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35452]) Branching to 'reflect-python3-5913'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

OK, ready for review. I started build:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/reflect-python3-5913

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. Argh `StringIO`/`BytesIO`.  If you think it's a good idea, perhaps `twisted.python.compat.NativeStringIO`?  Or we could wait until we need this check again, I guess, and refactor then.
  1. Explicit documentation about the return type of `safe_str` and `safe_repr` would probably be good at this point.
  1. The docstring for `unsignedID` says it is mostly only necessary before Python 2.5.  I guess keeping the existing implementation doesn't hurt, though.

Otherwise good.  Fix at least point 2 and then merge, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

  1. Opened [#5930](https://github.com/twisted/twisted/issues/5930).
  1. Done.
  1. Yeah, I noticed that too late. Plus probably Python 3.5 will make negative ids again, so can't hurt.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

(In [35463]) Branching to 'reflect-python3-5913-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set status to `closed`|
|-|-|

(In [35466]) Merge reflect-python3-5913-2

Author: itamar
Review: exarkun
Fixes: [#5913](https://github.com/twisted/twisted/issues/5913)

More functions from twisted.python.reflect ported to Python 3.
